server/src/services/media.service.ts

```diff
@@ -204,33 +224,50 @@ export class MediaService extends BaseService {
+    const processInvalidImages = process.env.IMMICH_PROCESS_INVALID_IMAGES === 'true';
+    const colorspace = this.isSRGB(asset) ? Colorspace.SRGB : image.colorspace;
+    const decodeOptions = { colorspace, processInvalidImages, size: image.preview.size };
+
+    let fullsizePath: string;
+    let extractedPath: string | undefined;
+    if (mimeTypes.isRaw(asset.originalPath)) {
+      let useExtracted = false;
+      extractedPath = StorageCore.getImagePath(asset, AssetPathType.EXTRACTED, ImageFormat.JPEG);
+      if (image.extractEmbedded) {
+        // try extracting embedded preview from RAW
+        // extracted image from RAW is always in JPEG format, as implied from the `jpgFromRaw` tag name
+        const didExtract = await this.mediaRepository.extract(asset.originalPath, extractedPath);
+        useExtracted = didExtract && (await this.shouldUseExtractedImage(extractedPath, image.preview.size));
+      }
 
-    try {
-      const useExtracted = didExtract && (await this.shouldUseExtractedImage(extractedPath, image.preview.size));
-      const inputPath = useExtracted ? extractedPath : asset.originalPath;
-      const colorspace = this.isSRGB(asset) ? Colorspace.SRGB : image.colorspace;
-      const processInvalidImages = process.env.IMMICH_PROCESS_INVALID_IMAGES === 'true';
-
-      const orientation = useExtracted && asset.exifInfo?.orientation ? Number(asset.exifInfo.orientation) : undefined;
-      const decodeOptions = { colorspace, processInvalidImages, size: image.preview.size, orientation };
-      const { data, info } = await this.mediaRepository.decodeImage(inputPath, decodeOptions);
-
-      const options = { colorspace, processInvalidImages, raw: info };
-      const outputs = await Promise.all([
-        this.mediaRepository.generateThumbnail(data, { ...image.thumbnail, ...options }, thumbnailPath),
-        this.mediaRepository.generateThumbnail(data, { ...image.preview, ...options }, previewPath),
-        this.mediaRepository.generateThumbhash(data, options),
-      ]);
-
-      return { previewPath, thumbnailPath, thumbhash: outputs[2] };
-    } finally {
-      if (didExtract) {
-        await this.storageRepository.unlink(extractedPath);
+      if (useExtracted) {
+        fullsizePath = extractedPath;
+      } else {
+        // did not extract or extracted preview is smaller than target size,
+        // convert a full-sized thumbnail from original instead
+        extractedPath = StorageCore.getImagePath(asset, AssetPathType.EXTRACTED, image.preview.format);
+        // const orientation = asset.exifInfo?.orientation ? Number(asset.exifInfo.orientation) : undefined;
+        await this.mediaRepository.generateThumbnail(
+          asset.originalPath,
+          { ...image.preview, colorspace, processInvalidImages, size: Infinity },
+          extractedPath,
+        );
       }
+      fullsizePath = extractedPath;
```

Comment:
Since this is set either way, it seems you can simply the if/else above it to a single `!useExtracted` check